File: Searching/Binary Search/remainder twist.java
import java.io.*;
import java.util.*;

public class Main {
    static class FastScanner {
        private final InputStream in;
        private final byte[] buffer = new byte[1 << 16];
        private int ptr = 0, len = 0;

        FastScanner(InputStream is) {
            in = is;
        }

        private int read() throws IOException {
            if (ptr >= len) {
                len = in.read(buffer);
                ptr = 0;
                if (len <= 0) return -1;
            }
            return buffer[ptr++];
        }

        long nextLong() throws IOException {
            long sign = 1, val = 0;
            int c;
            do {
                c = read();
            } while (c <= ' ');

            if (c == '-') {
                sign = -1;
                c = read();
            }

            while (c > ' ') {
                val = val * 10 + (c - '0');
                c = read();
            }
            return val * sign;
        }

        int nextInt() throws IOException {
            return (int) nextLong();
        }
    }

    static long countPairs(int n, int x) {
        if (x == 0) return 1L * n * n;

        long res = 0;

        for (int b = x + 1; b <= n; b++) {
            int q = n / b;
            int rem = n % b;

            res += 1L * q * (b - x);

            if (rem >= x) {
                res += (rem - x + 1);
            }
        }

        return res;
    }

    public static void main(String[] args) throws Exception {
        FastScanner fs = new FastScanner(System.in);
        StringBuilder out = new StringBuilder();

        int T = fs.nextInt();

        while (T-- > 0) {
            int N = fs.nextInt();
            long R = fs.nextLong();

            long total = 1L * N * N;

            if (R > total) {
                out.append(-1).append('\n');
                continue;
            }

            int lo = 0, hi = N - 1;
            int ans = 0;

            while (lo <= hi) {
                int mid = (lo + hi) >>> 1;

                if (countPairs(N, mid) >= R) {
                    ans = mid;
                    lo = mid + 1;
                } else {
                    hi = mid - 1;
                }
            }

            out.append(ans).append('\n');
        }

        System.out.print(out);
    }
}

File: Algorithms/src/hard/Remainder Twist.swift
import Foundation

final class FastScanner {
    private var data:[UInt8] = Array(FileHandle.standardInput.readDataToEndOfFile()) + [0]
    private var idx: Int = 0
    
    func readInt() -> Int {
        while data[idx] == 10 || data[idx] == 13 || data[idx] == 32 || data[idx] == 9 {
            idx += 1
        }
        
        var sign = 1
        if data[idx] == 45 {
            sign = -1
            idx += 1
        }
        
        var num = 0
        while data[idx] >= 48 && data[idx] <= 57 {
            num = num * 10 + Int(data[idx] - 48)
            idx += 1
        }
        
        return num * sign
    }
    
    func readLong() -> Int64 {
        while data[idx] == 10 || data[idx] == 13 || data[idx] == 32 || data[idx] == 9 {
            idx += 1
        }
        
        var sign: Int64 = 1
        if data[idx] == 45 {
            sign = -1
            idx += 1
        }
        
        var num: Int64 = 0
        while data[idx] >= 48 && data[idx] <= 57 {
            num = num * 10 + Int64(data[idx] - 48)
            idx += 1
        }
        
        return num * sign
    }
}

func countPairs(_ n: Int, _ x: Int) -> Int64 {
    if x == 0 {
        return Int64(n) * Int64(n)
    }
    
    var res: Int64 = 0
    
    if x + 1 > n {
        return 0
    }
    
    for b in (x + 1)...n {
        let q = n / b
        let rem = n % b
        
        res += Int64(q) * Int64(b - x)
        
        if rem >= x {
            res += Int64(rem - x + 1)
        }
    }
    
    return res
}

let fs = FastScanner()
let t = fs.readInt()

var answer = String()

for _ in 0..<t {
    let n = fs.readInt()
    let r = fs.readLong()
    
    let total = Int64(n) * Int64(n)
    
    if r > total {
        answer += "-1\n"
        continue
    }
    
    var lo = 0
    var hi = n - 1
    var ans = 0
    
    while lo <= hi {
        let mid = (lo + hi) >> 1
        
        if countPairs(n, mid) >= r {
            ans = mid
            lo = mid + 1
        } else {
            hi = mid - 1
        }
    }
    
    answer += "\(ans)\n"
}

print(answer, terminator: "")
